Added VStackExamples for Alignment and Spacing

SumanHansada opened the pull request:
This PR adds examples of VStack. The example includes alignment of VStack items (top, middle, bottom) as well as alignment of content inside VStack (leading, center, trailing). It also includes example of Scrollable VStack (when there is more content) as well as Scrollable LazyVStack (for efficient memory consumption in case of long items in VStack)

matthew-carroll commented:
@SumanHansada can you please add screenshots to this PR that show us the demos that you added?

SumanHansada commented:
Hi @matthew-carroll.. Added Screenshots!

| VStack | Image Preview | VStack | Image Preview |
|---|---|---|---|
| VStack Layout | <img width="200"  src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/1e8d54b9-2c95-4e3a-9578-96816e43de56" alt="VStack" > | Content Alignment | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/aaebca54-5374-4366-9579-c02fba8e765f" alt="Content Alignment" > |
| Top Aligned | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/6ece9cc9-3c30-4b8a-9ffc-a95d227f0c81" alt="Top Aligned" > | Content Spacing | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/1a3d591c-d05d-4323-b718-c31aafe149b1" alt="Content Spacing" > |
| Middle Aligned | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/eb80fd95-f24b-4cd0-bc69-681373cc9538" alt="Middle Aligned" > | VStack Exceeds Space | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/f53e3423-ef02-4789-8ce9-1c9d4972601c" alt="VStack Exceeds Space" > |
| Bottom Aligned | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/409e6312-c416-4250-9dd7-6fb5b70845d5" alt="Bottom Aligned" > | Lazy VStack Exceeds Space | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/80900e7a-7560-4e9f-b0a4-75286c11158d" alt="Lazy VStack Exceeds Space" > |

matthew-carroll commented:
@SumanHansada I'm a bit confused by what I'm seeing in the screenshots. The screenshots seem to mostly show lists. I can't tell which details are related to `VStack`s and which details are related to lists. Is there a reason that most of the examples are comprised of stuff that goes beyond `VStack`s? Can you reduce the examples so that they're pretty much just `VStack`s without bringing in other complexity?

I ask this because, as soon as we implement `VStack` in Flutter, we'll want to create the same demos that you created in Swift UI. But we can't create demos that are filled with Swift UI lists, because we're a long way away from implementing Swift UI lists.

SumanHansada commented:
Hi @matthew-carroll, I am updating changes.. will correct this and update ZStack, HStack by tomorrow or day after tomorrow at most. (Wednesday)

SumanHansada commented:
Hi @matthew-carroll .. I have added all the HStack and ZStack Examples.. Adding Screenshots as well.

matthew-carroll commented:
Hi @SumanHansada sorry I disappeared. It doesn't look like you responded to my earlier question about why the demos included lists. Did you remove the lists? If you changed the demos, can you please provide updated screenshots so I can see what the current demos look like?

SumanHansada commented:
Hi @matthew-carroll .. I had added list to make layouts more clear to understand and to reduce number of screens. Now I have removed the lists. I am adding all the screenshots.
# VStack
| Top Aligned | Middle Aligned | Bottom Aligned |
|:-:|:-:|:-:|
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/1e3dc7db-3e68-4297-8277-c1fdf7ce8dcc" alt="Top Aligned" > | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/3f22ae2c-88c5-4d4f-9215-c17c58c3c911" alt="Middle Aligned" > | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/6b0991f4-ce1c-4720-8f0f-b6aa290ccd56" alt="Bottom Aligned"> |
| Content Left Aligned | Content Center Aligned | Content Right Aligned |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/600d190a-45dc-40d3-b50f-393ca5efb74c" alt="Content Left Aligned" > | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/98a8cec9-6a9c-4fb7-a309-766899fa6a34" alt="Content Center Aligned"> | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/b2f41afd-50c9-433f-93fc-a784776e3060" alt="Content Right Aligned"> |
| Content Spacing | VStack Exceeds Space | Lazy VStack Exceeds Space |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/3fbcb41c-6401-4b80-9d07-9d78dfee15a6" alt="Content Spacing"> | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/6237f5bb-20c6-4ea0-9527-bd1dc1d1e761" alt="VStack Exceeds Space"> | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/5888d564-2191-45c5-8638-0a003ed1eb18" alt="Lazy VStack Exceeds Space"> |

SumanHansada commented:
# HStack
| Left Aligned | Center Aligned | Right Aligned |
| :-: | :-: | :-: |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/decfa603-8efc-402f-bb9e-5b2b2973285b" alt="Left Aligned">| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/f4ff9ebf-e8ec-4728-92c9-c9e9474f0742" alt="Center Aligned"> | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/6e419d7e-c376-4116-96dc-01d4c460a2c9" alt="Right Aligned"> |
| Content Top Aligned | Content Center Aligned | Content Bottom Aligned |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/c1557b86-1b86-4ace-8073-3a64cd559962" alt="Content Top Aligned" > |<img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/bf343eec-6b7e-483d-af82-40219baab6b7" alt="Content Center Aligned">| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/b059315f-b073-414d-b36e-7287a8f294eb" alt="Content Bottom Aligned"> |
| Content First Text Baseline | Content Last Text Baseline | HStack Exceeds Space |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/e2def939-a2b4-4779-aaf6-59ddae715f99" alt="Content First Text Baseline"> | <img  width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/76ed2f9e-2c95-4a2d-86ea-d0eaaf9caa6f" alt="Content Last Text Baseline">|<img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/abd52c6d-d187-4e3e-b97f-565e65447faa" alt="HStack Exceeds Space"> |
| Lazy HStack Exceeds Space |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/54bd92b6-4e13-4618-9191-802307adef05" alt="Lazy HStack Exceeds Space"> |

SumanHansada commented:
# ZStack
| ZStack Demo | ZStack Top Alignment | ZStack Top Leading Alignment |
| :-: | :-: | :-: |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/35004c7e-b28b-4c2e-9ef5-7fcc1566af7b" alt="ZStack Demo" >| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/8b932d9d-4c79-4440-9395-bfee6bd826fc" alt="ZStack Top Alignment">| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/90319a7f-5569-4c54-9e4b-6f9be422a309" alt="ZStack Top Leading Alignment" > |
| ZStack Top Trailing Alignment | ZStack Center Alignment | ZStack Bottom Alignment |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/78484446-a3ac-4107-898e-f1241d09a5d1" alt="ZStack Top Trailing Alignment" > | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/bfaa1c84-8be6-4099-8367-d6a32093a4c0" alt="ZStack Center Alignment" > | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/6ad77fee-5339-4034-b4ba-5a98965c1449" alt="ZStack Bottom Alignment"> |
| ZStack Bottom Leading Alignment | ZStack Bottom Trailing Alignment | |
| <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/0515819e-a2ca-4094-97af-615a0dd7e0b1" alt="ZStack Bottom Leading Alignment" > | <img width="200" src="https://github.com/Flutter-Bounty-Hunters/swift_ui/assets/9078433/a688bd5b-ba6e-4494-acc6-20c334159aa0" alt="ZStack Bottom Trailing Alignment" >| |

SumanHansada commented:
Hi @matthew-carroll .. Updated all the screenshots! 
Let me know if any more changes is required..

matthew-carroll commented on doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/layouts/HStackExamples.swift:
```diff
@@ -0,0 +1,227 @@
+                Text("First")
+                Text("Second")
+                Text("Third")
+                Spacer()
```
While this is a legitimate `HStack` example, it doesn't really reflect the name of "Left Aligned". This is an example of an `HStack` with an expanding child.

Isn't there an alignment property or alignment modifier that applies to `HStack`? If so, this example should work that way. The same goes for all other examples. Otherwise, all we're really testing is the movement of a `Spacer()` to different positions in the child list.

SumanHansada replied:
As per the Apple docs `HStack` uses `VerticalAlignment` and `VStack` uses `HorizontalAlignment`.
Ref: https://developer.apple.com/documentation/swiftui/aligning-views-within-a-stack#Customize-stack-alignment

So if we need to use `HStack` with `HorizontalAlignment`, one way is to wrap the `VStack` items inside `HStack` and then use it. It is also shown in the docs here: https://developer.apple.com/documentation/swiftui/horizontalalignment

Here in the above example individual items of `HStack` are `VStack` elements which have been given, (.leading, .center, .trailing) alignment. Other way is to use the `Spacer()`

matthew-carroll replied:
I'm not sure my point came across. You've labeled this example as "left aligned", however you're not really demonstrating alignment. You're demonstrating how a spacer works. A spacer can go anywhere, and the way a spacer works is different from how alignment works.

If you want to demonstrate `HStack` alignment, or `VStack` alignment, you should use the properties that are built into those views for alignment.

If you want to demonstrate how to use spacers, then the associated name should be about spacers, rather than alignment.

In this case, alignment is more important than spacers, so we should use the alignment properties, and keep the example names as-is.

matthew-carroll commented on doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/layouts/HStackExamples.swift:
```diff
@@ -0,0 +1,227 @@
+                Text("First").font(.largeTitle)
+                Text("Second").font(.title)
+                Text("Third").font(.title2)
+            }.border(Color.blue, width: 1)
```
What's the purpose for the `border()` in these examples? Is there a relationship between the definition of `HStack` and `border()`?

SumanHansada replied:
`border()` was added just to add visual clarity how the `HStack` content alignment looks like. Kind of used as reference. It is used in Apple docs as well. Mostly border or frames are used. Removed for now.
Ref - 
https://developer.apple.com/documentation/swiftui/verticalalignment
https://developer.apple.com/documentation/swiftui/horizontalalignment

matthew-carroll commented on doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/layouts/HStackExamples.swift:
```diff
@@ -0,0 +1,227 @@
+
+
+
+struct HStackLeftAlignedPage_Previews: PreviewProvider {
```
Let's place each of these preview configurations immediately below the view they apply to.

SumanHansada replied:
Done! Added previews just below the layouts.

matthew-carroll commented on doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/layouts/VStackExamples.swift:
```diff
@@ -2,66 +2,201 @@ import SwiftUI
+                Text("First")
+                Text("Second")
+                Text("Third")
+                Spacer()
```
Same comment here about the user of `Spacer`s

SumanHansada replied:
As per the Apple docs HStack uses VerticalAlignment and VStack uses HorizontalAlignment.
Ref: https://developer.apple.com/documentation/swiftui/aligning-views-within-a-stack#Customize-stack-alignment

So if we need to use HStack with HorizontalAlignment, one way is to wrap the VStack items inside HStack and then use it. It is also shown in the docs here: https://developer.apple.com/documentation/swiftui/horizontalalignment

Here in the above example individual items of HStack are VStack elements which have been given, (.leading, .center, .trailing) alignment. Other way is to use the Spacer()

matthew-carroll commented on doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/layouts/VStackExamples.swift:
```diff
@@ -2,66 +2,201 @@ import SwiftUI
+    }
+}
+
+struct VStackMiddleAlignedPage_Previews: PreviewProvider {
```
Let's move each of these previews below their respective views.

SumanHansada replied:
Done! Added previews just below the layouts.

matthew-carroll commented on doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/layouts/ZStackExamples.swift:
```diff
@@ -0,0 +1,155 @@
+        NavigationStack {
+            ZStack(alignment: .top) {
+                Rectangle().fill(.red).frame(width: 200, height: 200)
+                Rectangle().fill(.green).frame(width: /*@START_MENU_TOKEN@*/100/*@END_MENU_TOKEN@*/, height: 100)
```
What do "@START_MENU_TOKEN@" and "@END_MENU_TOKEN@" mean?

SumanHansada replied:
It was added by Xcode. Have removed it.

matthew-carroll commented on doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/layouts/ZStackExamples.swift:
```diff
@@ -0,0 +1,155 @@
+}
+
+
+struct ZStackDemo_Previews: PreviewProvider {
```
Let's move these previews beneath the view they apply to.

SumanHansada replied:
Done! Added previews just below the layouts.

matthew-carroll reviewed (commented):
Thanks for the updates. I left a few more comments.

SumanHansada commented:
@matthew-carroll, Sorry for the delay. Updated the code and added comments to the reviews. Let me know if any more changes are required.

matthew-carroll commented:
I'm currently implementing `VStack`, `HStack`, and `ZStack`. I'll add demos as needed. Thanks for trying to get the demos added.
